cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,420 @@
+			fwCache:             newFakeFirewallCache(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse: newFakeOKResponse(),
+			expectedGetResult:   newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedFirewall:    newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+		{
+			name:                 "return firewall from list when it matches the firewall name",
+			expectedFirewall:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			fwCache:              newFakeFirewallCache(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse:  newFakeNotFoundResponse(),
+			expectedListResponse: newFakeOKResponse(),
+			expectedListResult:   []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)},
+		},
+	}
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					getFunc: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedGetResult, test.expectedGetResponse, test.expectedGetError
+					},
+					listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+						return test.expectedListResult, test.expectedListResponse, test.expectedListError
+					},
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+
+			fw, err := fwManagerOp.Get(ctx)
+			if want, got := test.expectedError, err; !reflect.DeepEqual(want, got) {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+
+			if !cmp.Equal(fw, test.expectedFirewall) {
+				t.Errorf("got %q, want %q", fw, test.expectedFirewall)
+			}
+		})
+	}
+}
+
+func TestFirewallController_Set(t *testing.T) {
+	testcases := []struct {
+		name         string
+		fwCache      firewallCache
+		inboundRules []godo.InboundRule
+		firewall     *godo.Firewall
+		// Expected DO Firewall API, AddRules request, response values.
+		expectedAddRulesResponse *godo.Response
+		expectedAddRulesError    error
+		// Expected DO Firewall API, Create request, response values.
+		expectedCreateResult   *godo.Firewall
+		expectedCreateResponse *godo.Response
+		expectedCreateError    error
+		// Expected DO Firewall API, Get request, response values.
+		expectedGetResult   *godo.Firewall
+		expectedGetResponse *godo.Response
+		expectedGetError    error
+		expectedError       error
+		// Expected DO Firewall API, List request, response values.
+		expectedListResult   *godo.Firewall
+		expectedListResponse *godo.Response
+		expectedListError    error
+		// Expected DO Firewall API, Update request, response values.
+		expectedUpdateResult   *godo.Firewall
+		expectedUpdateResponse *godo.Response
+		expectedUpdateError    error
+	}{
+		{
+			name:                   "create firewall when cache does not exist (i.e. initial startup)",
+			fwCache:                newFakeFirewallCacheEmpty(),
+			expectedGetResponse:    newFakeNotFoundResponse(),
```

Comment:
AFAIU we should not call `Get()` on the firewall service for this particular test case; we'd straight go for the list instead since we do not have anything cached. Something I like to do to prevent false positives in test results (i.e., your test seems to pass but it actually only does so because it takes an unexpected code path) is to let unexpectedly called fakes/stubs fail hard or with a clear error message.

If you follow my suggestion from above to combine the return values from a godo service call into a single function, then we can leave that function stub for getting firewalls `nil` as we we do not expect the service call to be made. Thus, if it does get called unexpectedly, the test will panic hard, basically screaming in our face that we did something wrong.

**tl;dr:** only stub/fake/mock out relevant test doubles, and let everything else fail very clearly.